Add attribute parsing: Form&Render-Elements and Entity

n1kolas opened the pull request:
Hey,

stumbled on this project and think it's a great idea to have a language server for Drupal. This is already really helpful.
I was digging around in the code base and made some implementations.

**Attribute Parsing**
I've added attributes parsing and also changed the way parse_nodes work, because previously when a class_declaration was parsed, the containing attributes weren't parsed. That way the attributes in that class should also be caught by the match statement.

**Form Elements & Render Elements**
There are two new token types, that pick these up by their attributes.
I have not implemented any annotation based discovery, as those should be converted at some point anyways (https://www.drupal.org/project/drupal/issues/3396165).
Although I'm a bit unsure how the auto completion would work. Since they are both used within `['#type' => '...'`] it depends on the context (are we inside a Form f.e), but I don't think that's that easy to determine.
Maybe Render & Form Elements should always be suggested regardless of context?

**Auto generate snippets**
All picked up Form and Render Elements that have a `@code` and `@endcode` block inside their comment will be available as snippets with either `form-{id}` or `render-{id}`. Unfortunately a lot of these don't have usage examples in earlier versions and I don't think parsing the `getInfo()` method on the Plugin is a good idea. But this should improve with later versions (see screenshot).

**Entity Completion**
ContentEntityTypes are also picked up by attribute. Currently I suggest that completion when `getStorage()` by the entityTypeManager is used.

I hope this PR can help a bit, if not it was still fun to tinker with this a bit.
![form](https://github.com/user-attachments/assets/4a45ab40-577d-4aa2-9c3a-47094895b695)
![form2](https://github.com/user-attachments/assets/7015b3d2-2d9e-47f9-8e46-87557fe0da82)
![user](https://github.com/user-attachments/assets/7a0c207d-53bf-4fa3-b638-9cb43c28d04a)

jdrupal-dev commented:
Thank you for the contribution, this is a great addition!

I actually had a branch with an initial implementation on my machine, which I have just pushed: `feature/plugin-completion`.
That branch contains the parsing of attributes as well as completion of entity types, queue workers, field types, and data type plugins.

Maybe you can rebase the form and render implementation on top of that branch?

Thanks in advance 🙂

n1kolas commented:
@jdrupal-dev 
Thanks for the quick reply! That looks a lot cleaner.

I've updated the PR to target your feature branch and added the form- and render- snippet generation.

I also had to add some additional logic because Attributes that are defined without `id` for example: `#[RenderElement('date')]` weren't picked up.

e:/ Pushed a small code style improvement. Should be ready to review.

jdrupal-dev commented on src/parser/php.rs:
```diff
@@ -274,11 +274,26 @@ impl PhpParser {
                     class_attribute = Some(ClassAttribute::Plugin(DrupalPlugin {
                         plugin_type,
                         plugin_id,
+                        usage_example: None,
```
```suggestion
                        usage_example: self.extract_usage_example_from_comment(&comment_node),
```

n1kolas replied:
Fixed that.

jdrupal-dev commented on src/parser/php.rs:
```diff
@@ -330,6 +355,7 @@ impl PhpParser {
             Ok(plugin_type) => Some(ClassAttribute::Plugin(DrupalPlugin {
                 plugin_id,
                 plugin_type,
+                usage_example: None,
```
```suggestion
                usage_example: self.extract_usage_example_from_comment(&node.parent().prev_named_sibling()),
```
Maybe also check that node.parent().prev_named_sibling().kind() is equal to "comment".

n1kolas replied:
I check the node.kind() in the extract function now, which takes the comment_node argument.
If it's not a "comment" kind, I return None.

I had to move a couple of parent() calls up, since we were going from attribute -> attribute_group etc.

jdrupal-dev commented on src/parser/php.rs:
```diff
@@ -274,11 +274,26 @@ impl PhpParser {
             }
         }
 
+        // Try to get an usage example for snippet generation.
```
It would be nice if this was done in the context of building the class attribute. I have added two suggestions to how it could be done :)

n1kolas replied:
Sure, removed it and did it inline.

jdrupal-dev commented on src/server/handlers/completion/mod.rs:
```diff
@@ -326,10 +326,9 @@ pub fn handle_text_document_completion(request: Request) -> Option<Response> {
 fn get_global_snippets() -> Vec<CompletionItem> {
-    let mut snippets = HashMap::new();
-
+    let mut snippets: HashMap<String, String> = HashMap::new();
```
Is it not possible to use HashMap<&str, String> if you use .as_str() on the String types?

n1kolas replied:
I don't think so. I'm losing the fight against the borrow checker here with E0597 (value doesn't live long enough).

It works for the other inserts, since they are `&'static str` which live beyond the hashmap.
But &str's at runtime don't seem to live long enough.

Maybe you find a way, I haven't had any luck with it, although converting all the static's before to String does look a bit ugly..

n1kolas replied:
Oh, I've misread that. You only would convert the value to String. I'll check that.
e:/ Same issue for me. Couldn't find a way without using `HashMap<String, String>`

n1kolas replied:
I've found a way using`Box::leak` which converts them into a 'static str.
I pushed that, if you'd rather want `HashMap<String, String>` or another solution we can revert it.

https://doc.rust-lang.org/std/boxed/struct.Box.html#method.leak

jdrupal-dev replied:
I see what you were struggling with. I actually think it's best to use Strings since we anyway end up using a string in CompletionItem.
I have pushed a commit to your branch reverting to using strings.

jdrupal-dev commented on src/server/handlers/completion/mod.rs:
```diff
@@ -420,9 +437,44 @@ public static function create(ContainerInterface \$container, array \$configurat
+        })
+        .for_each(|(snippet_key_prefix, plugin_id, usage_example)| {
+            snippets.insert(
+                format!("{}-{}", snippet_key_prefix, plugin_id),
```
You can use format!(...).as_str() here instead of converting all the keys to Strings.

n1kolas replied:
See other comment on using &str vs String

jdrupal-dev reviewed (changes requested):
Looks very good, just a few small comments 😄

n1kolas commented:
Thanks for the review. Pushed some changes.

jdrupal-dev commented:
This looks good now. Merged!
